ruby/ql/lib/codeql/ruby/ast/Literal.qll

```diff
@@ -321,9 +322,16 @@ class StringInterpolationComponent extends StringComponent, StmtSequence,
 
   final override Stmt getStmt(int n) { toGenerated(result) = g.getChild(n) }
 
-  final override string getValueText() { none() }
-
   final override string getAPrimaryQlClass() { result = "StringInterpolationComponent" }
+
+  // If last statement in the interpolation is a constant or local variable read,
+  // we attempt to look up its string value.
+  // If there's a result, we return that as the string value of the interpolation.
+  final override string getValueText() {
```

Comment:
This (and the `getValueText` further down) should be implemented in `CfgNodes.qll` instead (i.e., on CFG nodes instead of AST nodes). In general, if `getValueText` is defined recursively, it should be done in `CfgNodes` for potential higher precision (in context of CFG splitting). Example
```rb
if (b)
  x = "true"
else
  x = "false"

if (b)
  puts x
```

If (when) Ruby has Boolean CFG splitting, the access to `x` in the `b = true` case will have constant value `"true"`, while it will have constant value `"false"` in the `b = false` case. However, the AST node `x` in `puts x` will not have a constant value.

Reply:
OK - I think I understand what you're saying here. I've updated this PR to push more calculation into the CFG. We still need a few changes at the AST level, I think, because regular string constants aren't CFG nodes. Let me know if I've got the right idea or if I'm still off.